packages/flutter_tools/lib/src/version.dart

```diff
@@ -829,20 +811,20 @@ class GitTagVersion {
   }
 
   String frameworkVersionFor(String revision) {
-    if (x == null || y == null || z == null || !revision.startsWith(hash)) {
+    if (x == null || y == null || z == null || (hash != null && !revision.startsWith(hash!))) {
       return '0.0.0-unknown';
```

Comment:
Actually the logic in this file seems wrong.  There are a few checks to fallback to other values when `frameworkVersion` is equal to `"unknown"` but that's not set anywhere, it's set to `0.0.0-unknown`.

https://github.com/flutter/flutter/blob/458f6f4513e9e943961d9df6d1a3cbee7c82a570/packages/flutter_tools/lib/src/version.dart#L158
https://github.com/flutter/flutter/blob/458f6f4513e9e943961d9df6d1a3cbee7c82a570/packages/flutter_tools/lib/src/version.dart#L273

Reply:
I fixed this up and added some tests.